src/HDF5.jl

```diff
@@ -1459,73 +1465,70 @@ function getindex(parent::Union{HDF5File, HDF5Group, HDF5Dataset}, r::HDF5Refere
-        elseif dtype<:HDF5.FixedArray && eltype(dtype)<:HDF5BitsKind
-            val = read!(io, Vector{eltype(dtype)}(undef,prod(size(dtype))))
-            push!(row, reshape(val, size(dtype)))
-        elseif dtype<:HDF5BitsKind
-            push!(row, read(io, dtype))
-        else
-            # for other types, just store the raw bytes and let the user
-            # decide what to do
-            push!(row, read!(io, Vector{UInt8}(undef,dsize)))
-        end
+# convert Cstring/FixedString to String
+function normalize_types(x::NamedTuple{T}) where T
+
+  vals = map(values(x)) do xi
+    Ti = typeof(xi)
+    if Ti == Cstring
+      unsafe_string(xi)
+    elseif Ti <: FixedString
+      join(Char.(xi.data))
+    elseif Ti <: FixedArray
+      reshape(collect(xi.data), size(Ti)...)
+    elseif Ti <: VariableArray
+      copy(unsafe_wrap(Array, convert(Ptr{eltype(xi)}, xi.p), xi.len, own=false))
+    elseif Ti <: NamedTuple
+      normalize_types(xi)
+    else
+      xi
     end
-    return (row...,)
+  end
+
+  NamedTuple{T}(vals)
 end
 
-# Read compound type
-function read(obj::HDF5Dataset, T::Union{Type{Array{HDF5Compound{N}}},Type{HDF5Compound{N}}}) where {N}
-    t = datatype(obj)
-    local sz = 0; local n;
-    local membername; local membertype;
-    local memberoffset; local memberfiletype; local membersize;
-    try
-        memberfiletype = Vector{HDF5Datatype}(undef,N)
-        membertype = Vector{Type}(undef,N)
-        membername = Vector{String}(undef,N)
-        memberoffset = Vector{UInt64}(undef,N)
-        membersize = Vector{UInt32}(undef,N)
-        for i = 1:N
-            filetype = HDF5Datatype(h5t_get_member_type(t.id, i-1))
-            memberfiletype[i] = filetype
-            membertype[i] = hdf5_to_julia_eltype(filetype)
-            memberoffset[i] = sz
-            membersize[i] = sizeof(filetype)
-            sz += sizeof(filetype)
-            membername[i] = h5t_get_member_name(t.id, i-1)
-        end
-    finally
-        close(t)
-    end
-    # Build the "memory type"
-    memtype_id = h5t_create(H5T_COMPOUND, sz)
-    for i = 1:N
-        h5t_insert(memtype_id, membername[i], memberoffset[i], memberfiletype[i].id) # FIXME strings
-    end
-    # Read the raw data
-    buf = Vector{UInt8}(undef,length(obj)*sz)
-    h5d_read(obj.id, memtype_id, H5S_ALL, H5S_ALL, obj.xfer, buf)
-
-    # Convert to the appropriate data format using iobuffer
-    iobuff = IOBuffer(buf)
-    data = Any[]
-    while !eof(iobuff)
-        push!(data, read_row(iobuff, membertype, membersize))
-    end
-    # convert HDF5Compound type parameters to tuples
-    membername = (membername...,)
-    membertype = (membertype...,)
-    if T === HDF5Compound{N}
-        return HDF5Compound(data[1], membername, membertype)
+# get a vector of all the leaf types in a (possibly nested) named tuple
+function get_all_types(::Type{NamedTuple{T, U}}) where T where U
```

Comment:
I'm guessing there's a better way to do this that I can't currently think of, but seems fine for now.

Reply:
This seems faster / better

```julia
function get_all_types1(::Type{NamedTuple})
	types = DataType[]
	for Ui in U.types
	    if Ui <: NamedTuple
	      append!(types, get_all_types(Ui))
	    else
	      push!(types, Ui)
	    end
	  end
	  return types
end
```

Reply:
maybe this would be better/more julian?

```
do_normalize(::Type{T}) where T = false
do_normalize(::Type{NamedTuple{T, U}}) where T where U = any(i -> do_normalize(fieldtype(U,i)), 1:fieldcount(U))
do_normalize(::Type{T}) where T <: Union{Cstring, FixedString, FixedArray, VariableArray} = true

do_reclaim(::Type{T}) where T = false
do_reclaim(::Type{NamedTuple{T, U}}) where T where U =  any(i -> do_reclaim(fieldtype(U,i)), 1:fieldcount(U))
do_reclaim(::Type{T}) where T <: Union{Cstring, VariableArray} = true
```

the only reason I collect all the types is to see if I need to call reclaim or do an extra normalization step to turn the data into a native julia type.

Reply:
yeah that might be simpler. It's certainly in our best favor to make the code run as fast as possible and be as simple as possible to read for maintenance.  You can also benchmark performance.